internal/pkg/gateway/api.go

```diff
@@ -66,13 +72,28 @@ func (gs *Server) Endorse(ctx context.Context, proposedTransaction *gp.ProposedT
 	}
 
 	var responses []*peer.ProposalResponse
-	// send to all the endorsers - TODO fan out in parallel
+	// send to all the endorsers
+	c := make(chan *endorserResponse)
 	for _, endorser := range endorsers {
-		response, err := endorser.ProcessProposal(ctx, signedProposal)
-		if err != nil {
-			return nil, fmt.Errorf("failed to process proposal: %w", err)
+		go func(endorser peer.EndorserClient) {
+			response, err := endorser.ProcessProposal(ctx, signedProposal)
+			c <- &endorserResponse{pr: response, err: err}
+		}(endorser)
+	}
+
+	timeout := time.After(gs.options.EndorsementTimeout)
```

Comment:
Under load, this can result in significant memory pressure. This problem is highlighted in the doc: `The underlying Timer is not recovered by the garbage collector until the timer fires`

That said, why do this? Why not decorate the context with a timeout and use the same one across all of the calls? (I'm guessing you copied this code from somewhere...) Once you know the calls have a proper timeout and cancelation, you could choose to keep using a select in a for loop or move to something simpler (like a WaitGroup) to wait for the calls to complete.

Reply:
rewritten to use context.WithTimeout, as suggested